rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+### Configuration
+
+We will make the following changes to `DbOptions`:
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// the Clock to use for insertion timestamps
+    ///
+    /// default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
+
+    /// the default time-to-live (TTL) for insertions (note that re-inserting a key
+    /// with any value will update the TTL to use the default_ttl)
+    ///
+    /// default: no TTL (insertions will remain until deleted)
+    default_ttl: Option<Duration>
+}
+```
+
+`Db#put_put_with_options` can optionally specify a timestamp and a row-level TTL at insertion time 
+by leveraging newly added fields in `WriteOptions`:
+
+```rust
+pub struct WriteOptions {
+    // ...
+   
+    /// the timestamp (in millis) at which this operation is considered to be executed - 
+    /// note that SlateDB does not automatically resolve out-of-order insertions
+    /// (if you need to retrieve the most recent semantic insertion, specify the
+    /// option on ReadOptions)
+    ///
+    /// default: uses the clock time configured in the DbOptions
+    pub timestamp_ms: Option<i64>,
+   
+    /// the time-to-live (ttl) for this insertion. If this insert overwrites an existing
+    /// database entry, the TTL for the most recent entry will be canonical.
+    ///
+    /// default: the TTL configured in DbOptions when opening a SlateDB session
+    pub ttl: Option<Duration>,
+}
+```
+
+`Db#get_with_options` can optionally specify: a grace period with which to continue searching the
+LSM tree for a more recent insertion. This allows you to specify an upper bound for what you expect
+out-of-order insertions to be.
+
+```rust
+pub struct ReadOptions {
+    // ...
+
+    /// if set to a positive number, SlateDB will not return the first matching row for
+    /// each key lookup and instead will continue to traverse the LSM tree until the next
+    /// matching key is older than the found key's timestamp + retrieval_grace_ms. Logically,
+    /// this is an "upper bound" on the magnitude of out-of-order insertions.
+    ///
+    /// Example: if there are three insertions [KeyA@100, KeyA@50, KeyA@150] that are each
```

Comment:
Another question: will this work with caching? Need to noodle on this. How will it behave if KeyA@100 vs. KeyA@50 vs. KeyA@150 is in the in-memory cache? I guess since we're caching based on blocks, it should be OK (since the blocks are attached to an SST).

Reply:
Not sure I see how caching changes this? If you issue a read with a `grace_ms > 0` you always have to attempt to find at least one "deeper" key.